tests/e2e/tests.bats

```diff
@@ -9,6 +9,31 @@ assertGenre() {
 	)
 }
 
+tempConfigWithAutoEnabled() {
+	TMP_CONF=/tmp/beets-autogenre-test-config.yaml
+	cp -f /etc/beets/default-config.yaml $TMP_CONF
+	printf '\nautogenre:\n  auto: true\n' >> $TMP_CONF
+}
+
+@test 'auto-detect song genre on import' {
+	# TODO: fix failing logic. fails with "sqlite3.OperationalError: no such column: genres"
+	# Tuba Skinny - Jubilee Stomp
+	tempConfigWithAutoEnabled
+	beet -c $TMP_CONF ytimport -q --quiet-fallback=asis https://www.youtube.com/watch?v=jft3BVoxqjo
+	QUERY='Jubilee Stomp'
+	assertGenre "$QUERY" 'lastfm | Jazz | Jazz'
+}
```

Comment:
Unfortunately, the new test cases for auto-detecting the genre on import fail because the `genres` column does not exist:
```
not ok 1 auto-detect song genre on import in 8268ms
# (in test file tests/e2e/tests.bats, line 22)
#   `beet -c $TMP_CONF ytimport -q --quiet-fallback=asis https://www.youtube.com/watch?v=jft3BVoxqjo' failed
# ytimport: Ignoring 0 known URLs
# ytimport: Downloading 1 song(s) to /data/ytimport
# [youtube] Extracting URL: https://www.youtube.com/watch?v=jft3BVoxqjo
# [youtube] jft3BVoxqjo: Downloading webpage
# [youtube] jft3BVoxqjo: Downloading ios player API JSON
# [youtube] jft3BVoxqjo: Downloading android player API JSON
# WARNING: [youtube] Skipping player responses from android clients (got player responses for video "aQvGIIdgFDM" instead of "jft3BVoxqjo")
# [youtube] jft3BVoxqjo: Downloading player 17fd9675
# WARNING: Writing cache to '/.cache/yt-dlp/youtube-sigfuncs/js_17fd9675_109.json' failed: Traceback (most recent call last):
#   File "/usr/local/lib/python3.12/site-packages/yt_dlp/cache.py", line 41, in store
#     os.makedirs(os.path.dirname(fn), exist_ok=True)
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 225, in makedirs
# PermissionError: [Errno 13] Permission denied: '/.cache'
#
# WARNING: Writing cache to '/.cache/yt-dlp/youtube-nsig/17fd9675.json' failed: Traceback (most recent call last):
#   File "/usr/local/lib/python3.12/site-packages/yt_dlp/cache.py", line 41, in store
#     os.makedirs(os.path.dirname(fn), exist_ok=True)
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 225, in makedirs
# PermissionError: [Errno 13] Permission denied: '/.cache'
#
# WARNING: Writing cache to '/.cache/yt-dlp/youtube-nsig/17fd9675.json' failed: Traceback (most recent call last):
#   File "/usr/local/lib/python3.12/site-packages/yt_dlp/cache.py", line 41, in store
#     os.makedirs(os.path.dirname(fn), exist_ok=True)
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 225, in makedirs
# PermissionError: [Errno 13] Permission denied: '/.cache'
#
# WARNING: Writing cache to '/.cache/yt-dlp/youtube-sigfuncs/js_17fd9675_105.json' failed: Traceback (most recent call last):
#   File "/usr/local/lib/python3.12/site-packages/yt_dlp/cache.py", line 41, in store
#     os.makedirs(os.path.dirname(fn), exist_ok=True)
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 215, in makedirs
#   File "<frozen os>", line 225, in makedirs
# PermissionError: [Errno 13] Permission denied: '/.cache'
#
# [youtube] jft3BVoxqjo: Downloading m3u8 information
# [info] jft3BVoxqjo: Downloading 1 format(s): 251
# [info] Downloading video thumbnail 44 ...
# [info] Writing video thumbnail 44 to: /data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].webp
# [download] Destination: /data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].webm
# 
[download]   0.0% of    2.99MiB at  655.97KiB/s ETA 00:04
[download]   0.1% of    2.99MiB at    1.55MiB/s ETA 00:01
[download]   0.2% of    2.99MiB at    3.13MiB/s ETA 00:00
[download]   0.5% of    2.99MiB at    5.88MiB/s ETA 00:00
[download]   1.0% of    2.99MiB at    3.82MiB/s ETA 00:00
[download]   2.1% of    2.99MiB at    4.01MiB/s ETA 00:00
[download]   4.1% of    2.99MiB at    5.60MiB/s ETA 00:00
[download]   8.3% of    2.99MiB at    8.26MiB/s ETA 00:00
[download]  16.7% of    2.99MiB at   13.00MiB/s ETA 00:00
[download]  33.4% of    2.99MiB at   20.60MiB/s ETA 00:00
[download]  66.8% of    2.99MiB at   34.83MiB/s ETA 00:00
[download] 100.0% of    2.99MiB at   47.27MiB/s ETA 00:00
[download] 100% of    2.99MiB in 00:00:00 at 27.43MiB/s
# [ExtractAudio] Destination: /data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].opus
# Deleting original file /data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].webm (pass -k to keep)
# [MetadataFromField] Parsed meta_comment from '%(webpage_url)s %(title)s': 'https://www.youtube.com/watch?v=jft3BVoxqjo Tuba Skinny - Jubilee Stomp - Royal Street I'
# [MetadataFromField] Parsed artist from '%(uploader)s': 'TheWsm'
# [MetadataFromField] Parsed artist from '%(title)s': 'Tuba Skinny'
# [MetadataFromField] Parsed title from '%(title)s': 'Jubilee Stomp - Royal Street I'
# [MetadataFromField] Parsed album_artist from '%(artist)s': 'Tuba Skinny'
# [MetadataFromField] Parsed meta_yt_id from '%(id)s': 'jft3BVoxqjo'
# [MetadataFromField] Parsed meta_yt_source from '%(webpage_url_domain)s': 'youtube.com'
# [MetadataFromField] Parsed meta_yt_source from '%(webpage_url_domain)s': 'youtube'
# [MetadataFromField] Parsed meta_yt_likes from '%(like_count)s': '175404'
# [MetadataFromField] Parsed meta_yt_dislikes from '%(dislike_count)s': 'NA'
# [MetadataFromField] Parsed meta_yt_views from '%(view_count)s': '8131691'
# [MetadataFromField] Parsed meta_yt_rating from '%(average_rating)s': 'NA'
# [MetadataFromField] Parsed meta_publish_date from '%(release_date>%Y-%m-%d,upload_date>%Y-%m-%d)s': '2018-04-25'
# [Metadata] Adding metadata to "/data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].opus"
# [ThumbnailsConvertor] Converting thumbnail "/data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].webp" to png
# [EmbedThumbnail] mutagen: Adding thumbnail to "/data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].opus"
# [XAttrMetadata] Writing metadata to file's xattrs
# [Rename] Renaming '/data/ytimport/singles/Tuba_Skinny_-_Jubilee_Stomp_-_Royal_Street_I [jft3BVoxqjo].opus' to '/data/ytimport/singles/Tuba Skinny - Jubilee Stomp - Royal Street I [jft3BVoxqjo].opus'
# [SplitChaptersToTracks] Splitting /data/ytimport/singles/Tuba Skinny - Jubilee Stomp - Royal Street I [jft3BVoxqjo].opus
# ytimport: Importing downloaded songs into beets library
# No files imported from /data/ytimport/albums
#
# /data/ytimport/singles/Tuba Skinny - Jubilee Stomp - Royal Street I [jft3BVoxqjo].opus
# Importing as-is.
# Traceback (most recent call last):
#   File "/usr/local/bin/beet", line 8, in <module>
#     sys.exit(main())
#              ^^^^^^
#   File "/usr/local/lib/python3.12/site-packages/beets/ui/__init__.py", line 1865, in main
#     _raw_main(args)
#   File "/usr/local/lib/python3.12/site-packages/beets/ui/__init__.py", line 1852, in _raw_main
#     subcommand.func(lib, suboptions, subargs)
#   File "/usr/local/lib/python3.12/site-packages/beetsplug/ytimport/__init__.py", line 101, in run_import_cmd
#     self._import_files(lib, opts, singles_dir, True)
#   File "/usr/local/lib/python3.12/site-packages/beetsplug/ytimport/__init__.py", line 202, in _import_files
#     import_files(lib, [src_dir], None)
#   File "/usr/local/lib/python3.12/site-packages/beets/ui/commands.py", line 1326, in import_files
#     session.run()
#   File "/usr/local/lib/python3.12/site-packages/beets/importer.py", line 360, in run
#     pl.run_parallel(QUEUE_SIZE)
#   File "/usr/local/lib/python3.12/site-packages/beets/util/pipeline.py", line 447, in run_parallel
#     raise exc_info[1].with_traceback(exc_info[2])
#   File "/usr/local/lib/python3.12/site-packages/beets/util/pipeline.py", line 312, in run
#     out = self.coro.send(msg)
#           ^^^^^^^^^^^^^^^^^^^
#   File "/usr/local/lib/python3.12/site-packages/beets/util/pipeline.py", line 195, in coro
#     func(*(args + (task,)))
#   File "/usr/local/lib/python3.12/site-packages/beets/importer.py", line 1661, in plugin_stage
#     func(session, task)
#   File "/usr/local/lib/python3.12/site-packages/beets/plugins.py", line 143, in wrapper
#     return func(*args, **kwargs)
#            ^^^^^^^^^^^^^^^^^^^^^
#   File "/usr/local/lib/python3.12/site-packages/beetsplug/autogenre/__init__.py", line 74, in imported
#     self._update_item_genre(task.item)
#   File "/usr/local/lib/python3.12/site-packages/beetsplug/autogenre/__init__.py", line 85, in _update_item_genre
#     item.store(['genre', 'genres', 'genre_source'])
#   File "/usr/local/lib/python3.12/site-packages/beets/library.py", line 395, in store
#     super().store(fields)
#   File "/usr/local/lib/python3.12/site-packages/beets/dbcore/db.py", line 585, in store
#     tx.mutate(query, subvars)
#   File "/usr/local/lib/python3.12/site-packages/beets/dbcore/db.py", line 962, in mutate
#     cursor = self.db._connection().execute(statement, subvals)
#              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
# sqlite3.OperationalError: no such column: genres
```

Though, that is unexpected since the same code works when I invoke it using the `beet autogenre` command and that command is also not explicitly creating that field.
@JOJ0 can you tell me by any chance whether this is a Beets bug or expected behaviour and if so how to fix it?

Reply:
![image](https://github.com/mgoltzsche/beets-autogenre/assets/2733783/ec03e1b9-8554-4b80-ba82-ef7042324257)

I'm sorry iphone app is making me crazy, cant copy line links... 😵‍💫

somewhere here you could do some low level debug prints.

one thing I'm not sure about: You are importing "asis". that skips quite some things in the beets import process. Maybe there are things like properly handling flexible attributes that don't run?

have you tried running a simple beets asis import and adding

--set=randomname=randomvalue 

to create a flex field does that work?

Reply:
I'm sorry I was mistaken, forget about the asis import idea, you are ytimporting and falling back to a genre named asis. Ok that shouldn't be a problem

Reply:
maybe also poking around here with debug prints brings you closer to understanding whats happening:

![image](https://github.com/mgoltzsche/beets-autogenre/assets/2733783/c8fca4df-8e8c-4439-83e6-9484db78c36e)

Reply:
> I'm sorry I was mistaken, forget about the asis import idea, you are ytimporting and falling back to a genre named asis. Ok that shouldn't be a problem

I'm sorry I shouldnt write anything before coffee. That was wrong again.

You are using https://github.com/mgoltzsche/beets-ytimport to import which will probably pass through the quiet fallback option to an actual beets command, right? (Haven't actually read what ytimport does, but it looks like you built a wrapper to beets in Pyrhon)

Reply:
Sorry, I missed to mention that I use the `ytimport` plugin to load test data within those integration tests (to make them easily reproducible) and that the plugin actually just [calls Beet's `import_files` function](https://github.com/mgoltzsche/beets-ytimport/blob/v1.8.1/beetsplug/ytimport/__init__.py#L202) after downloading the given URL as an opus audio file to `ytimport/singles`. Alternatively, the error can be reproduced using the `beet import` command directly, showing that it is not related to the `ytimport` plugin:
```
beet ytimport --no-import https://www.youtube.com/watch?v=jft3BVoxqjo
beet import -qs --quiet-fallback=asis ytimport/singles/
```

Setting a flex attribute during import works (when the autogenre plugin's auto option is disabled):
```
$ beet import -qs --quiet-fallback=asis --set=mykey=myvalue ytimport/singles/
$ beet ls -f '$mykey'
myvalue
```

Similarly, when I import via the `ytimport` plugin (with autogenre plugin's auto option disabled), it sets additional flex attributes (e.g. `yt_id`) by default successfully. However, as opposed to the autogenre plugin, the ytimport plugin explicitly registers additional media fields by [calling the `self.add_media_field` function](https://github.com/mgoltzsche/beets-ytimport/blob/v1.8.1/beetsplug/ytimport/__init__.py#L39-L43), in case that matters.
The autogenre plugin does not define the `genres` field itself because it is [already defined by default](https://github.com/beetbox/mediafile/blob/6accdc4c4a7fe4d808b674a97fb5474e12621504/mediafile.py#L1808) - but apparently not known within the DB schema.
Though, I remember at some point I had removed the `genres` field usage from the code completely, leaving the `genre_source` field as an additional media field that should end up as flex attribute but then the item write operation failed again - but this time with `sqlite3.OperationalError: no such column: genre_source`.

Reply:
> The error points to beets trying to access it via an actual sql statement, which is just wrong. somewhere higher in the code it should realize that this is not a fixed db field and only a flex attr.

As you said in the other issue. That might be the case. I'll have to continue to look into that and the DB schema especially...

Reply:
Ah, saving the regular item fields separate from the flex attributes made it work:
Using 
```
item.store(['genre'])
item.store(['genres', 'genre_source'])
```
instead of what I had previously:
```
item.store(['genre', 'genres', 'genre_source'])
```

Reply:
Thanks for your help pointing out that the mix of regular and flex attributes might be the problem!